internal/protoc/compiler.go

```diff
@@ -0,0 +1,736 @@
+		// if not an exec.ExitError, short circuit
+		if _, ok := runErr.(*exec.ExitError); !ok {
+			return nil, runErr
+		}
+	}
+	output := strings.TrimSpace(buffer.String())
+	if output != "" {
+		c.logger.Debug("protoc output", zap.String("output", output))
+	}
+	failures, err := parseProtocOutput(cmdMeta, output)
+	if err != nil {
+		return nil, err
+	}
+	// we had a run error but for whatever reason did not get any parsed
+	// output lines, we still want to fail in this case
+	// this generally should not happen, especially as plugins that fail
+	// will result in a pluginFailedRegexp matching line but this
+	// is just to make sure
+	if len(failures) == 0 && runErr != nil {
+		return nil, runErr
+	}
+	return failures, nil
+}
+
+func (c *compiler) getCmdMetas(protoSet *file.ProtoSet) (cmdMetas []*cmdMeta, retErr error) {
+	defer func() {
+		// if we error in this function, we clean ourselves up
+		if retErr != nil {
+			cleanCmdMetas(cmdMetas)
+			cmdMetas = nil
+		}
+	}()
+	// you need a new downloader for every ProtoSet as each prototool.yaml could
+	// have a different protoc_version value
+	downloader := c.newDownloader(protoSet.Config)
+	if _, err := downloader.Download(); err != nil {
+		return cmdMetas, err
+	}
+	for dirPath, protoFiles := range protoSet.DirPathToFiles {
+		// you want your proto files to be in at least one of the -I directories
+		// or otherwise things can get weird
+		// we make best effort to make sure we have the a parent directory of the file
+		// if we have a config, use that directory, otherwise use the working directory
+		configDirPath := protoSet.Config.DirPath
+		if configDirPath == "" {
+			configDirPath = protoSet.WorkDirPath
+		}
+		includes, err := getIncludes(downloader, protoSet.Config, dirPath, configDirPath)
+		if err != nil {
+			return cmdMetas, err
+		}
+		var args []string
+		for _, include := range includes {
+			args = append(args, "-I", include)
+		}
+		protocPath, err := downloader.ProtocPath()
+		if err != nil {
+			return cmdMetas, err
+		}
+		// this could really use some refactoring
+		// descriptorSetFilePath will either be a temporary file that we output
+		// a file descriptor set to, or the system equivalent of /dev/null
+		// isTempFile is effectively != /dev/null for all intents and purposes
+		// we do -o /dev/null because protoc needs at least one output, but in the compile-only
+		// mode, we want to just test for compile failures
+		descriptorSetFilePath, isTempFile, err := c.getDescriptorSetFilePath(protoSet)
+		if err != nil {
+			return cmdMetas, err
+		}
+		if descriptorSetFilePath != "" {
+			descriptorSetTempFilePath := descriptorSetFilePath
+			if !isTempFile {
+				descriptorSetTempFilePath = ""
+			}
+			// either /dev/null or a temporary file
+			iArgs := append(args, "-o", descriptorSetFilePath)
+			// if its a temporary file, that means we actually care about the output
+			// so we do --include_imports to get all necessary info in the output file descriptor set
```

Comment:
You can in theory, it's just not what most plugins expect (and I'm nervous about passing a FileDescriptorSet with imports to golang/protobuf/protoc-gen-go, for example, I think this will break golang/protobuf/protoc-gen-go). All this would simplify is this if statement. Perhaps it could be better explained, any thoughts?

Reply:
same here, we can revisit this after